pkg/controllers/user/pipeline/controller/pipeline/cronsyncer.go

```diff
@@ -0,0 +1,116 @@
+package pipeline
+
+import (
+	"context"
+	"github.com/pkg/errors"
+	"github.com/rancher/rancher/pkg/controllers/user/pipeline/utils"
+	"github.com/rancher/rancher/pkg/ticker"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/robfig/cron"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/labels"
+	"time"
+)
+
+const (
+	syncInterval = 60 * time.Second
+)
+
+type CronSyncer struct {
+	pipelineLister        v3.PipelineLister
+	pipelines             v3.PipelineInterface
+	pipelineExecutions    v3.PipelineExecutionInterface
+	pipelienExecutionLogs v3.PipelineExecutionLogInterface
+}
+
+func (s *CronSyncer) sync(ctx context.Context, syncInterval time.Duration) {
+	for range ticker.Context(ctx, syncInterval) {
+		logrus.Debugf("Start sync pipeline cron")
+		s.syncCron()
+		logrus.Debugf("Sync pipeline cron complete")
+	}
+}
+
+func (s *CronSyncer) syncCron() {
+	pipelines, err := s.pipelineLister.List("", labels.NewSelector())
+	if err != nil {
+		logrus.Errorf("Error listing pipelines")
+		return
+	}
+	for _, p := range pipelines {
+		s.checkCron(p)
+	}
+}
+
+func (s *CronSyncer) checkCron(pipeline *v3.Pipeline) {
+	if pipeline.Spec.TriggerCronExpression == "" {
+		return
+	}
+	if pipeline.Status.NextStart == "" {
+		//update nextstart time
+		nextStart, err := getNextStartTime(pipeline.Spec.TriggerCronExpression, pipeline.Spec.TriggerCronTimezone, time.Now())
+		if err != nil {
+			logrus.Errorf("Error getNextStartTime - %v", err)
+			return
+		}
+		pipeline.Status.NextStart = nextStart
+		if _, err := s.pipelines.Update(pipeline); err != nil {
+			logrus.Errorf("Error update pipeline - %v", err)
+		}
+		return
+	}
+
+	nextStartTime, err := time.ParseInLocation(time.RFC3339, pipeline.Status.NextStart, time.Local)
+	if err != nil {
+		logrus.Errorf("Error parsing nextStart - %v", err)
+		s.resetNextRun(pipeline)
+		return
+	}
+	if nextStartTime.After(time.Now()) {
+		//not time up
+		return
+	} else if nextStartTime.Before(time.Now()) && nextStartTime.Add(syncInterval).After(time.Now()) {
```

Comment:
It might be. I think this is less overhead than keeping cron go routines running for every pipeline and ensure they are on time. 
To relieve the miss, I think I will 
1. only loop pipelines with cron trigger
2. log error if trigger time goes stale.

Reply:
If the change takes too much time, given we have to merge it before Weds, maybe we can do it latter in a separate PR